### DISCUSSION TITLE
Configuring application objects without invoking run()

### DISCUSSION THREAD

**Discussion by v-p-b** [ajalt/clikt#564] [Q&A]
I have a project that uses Clikt for command line parsing, but I also want to use it as a library. In the latter case I don't want the main entry point of my program to be executed, but I want to expose API's (called by my entry point when used from the cmdline) for my caller. It'd make sense to me to also use Clikt to configure my class instance, but when implementing tests I found that `parse()` implicitly invokes `run()`, messing up the state of my object (I'm using iterators internally...):

```kotlin
val myCliktApp = MyCliktApp()
myCliktApp.apply{parse(argv)}
```

The only workaround I found is to introduce an extra ("-A - You are an API now") option, that makes `run()` exit early. I still wonder if there's a way to configure the object (set its members) via Clikt, without invoking `run()`?

**Comment by ajalt** [comment]
Sure. `parse` is just a [one line convenience function][1], so you can use the `CommandLineParser` functions directly, for example:

```kotlin
CommandLineParser.parseAndRun(myCliktApp, argv) { /* this is empty instead of calling it.run() */ }
```

There are also other ways to [test your code][2] that you might want instead.

[1]: https://github.com/ajalt/clikt/blob/5593cd0b3645a2862b45ea574b1ae0dcb2e5db91/clikt/src/commonMain/kotlin/com/github/ajalt/clikt/core/CoreCliktCommand.kt#L107
[2]: https://ajalt.github.io/clikt/testing/

**Comment by v-p-b** [comment]
Thank you this works like a charm!

The testing docs in part convinced me that Clikt will worth using in this project, but my test setup turned out to be a bit more convoluted than standard user-cases :)
